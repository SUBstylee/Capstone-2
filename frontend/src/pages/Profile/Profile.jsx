import { useState,useEffect } from "react";
import { useNavigate } from "react-router-dom";
import {Form,Button,Row,Col,Table} from 'react-bootstrap';
import {LinkContainer} from 'react-router-bootstrap';
import { useDispatch,useSelector } from "react-redux";
import Message from "../../components/Message/Message";
import Loader from '../../components/Loader/Loader';
import CustomInput from "../../components/CustomInput/CustomInput";
import {getUserDetails,updateUserProfile} from '../../actions/userActions';
import {listMyOrders} from '../../actions/orderActions';
import { USER_UPDATE_PROFILE_RESET } from "../../constants/userConstants";

const Profile = () => {
    const [name,setName]=useState('')
    const [email,setEmail]=useState('');
    const [password,setPassword]=useState('');
    const [confirmPassword,setConfirmPassword]=useState('');
    const [message,setMessage]=useState(null);

    const navigate = useNavigate();
    const dispatch=useDispatch();
    const userDetails=useSelector((state)=>state.userDetails);
    const {loading,error,user}=userDetails;
    const userLogin=useSelector((state)=>state.userLogin);
    const {userInfo}=userLogin;
    const userUpdateProfile=useSelector((state)=>state.userUpdateProfile);
    const {success}=userUpdateProfile;
    const myOrderList=useSelector((state)=>state.orderListMyOrders);
    const {loading:loadingOrders,error:errorOrders,orders}=myOrderList;

    useEffect(()=>{
        if(!userInfo){
            navigate('/login');
        }else{
            if(!user||!user.name||success){
                dispatch({type:USER_UPDATE_PROFILE_RESET});
                dispatch(getUserDetails('profile'));
                dispatch(listMyOrders());
            }else{
                setName(user.name);
                setEmail(user.email);
            };
        };
    },[navigate,userInfo,dispatch,user,success]);
    
    const submitHandler=(e)=>{
        e.preventDefault();
        if(password!==confirmPassword){
            setMessage('Passwords do not match!')
        }else{
            dispatch(updateUserProfile({id:user._id,name,email,password}))
        };
        
    };

  return (
    <div className="Profile">
        {loading&&<Loader/>}
        <Row>
            <Col lg={4}>
                <h2>User Profile</h2>
                {message&&<Message variant='danger'>{message}</Message>}
                {error&&<Message variant='danger'>{error}</Message>}
                {success&&<Message variant='success'>Profile Updated!</Message>}
                <Form onSubmit={submitHandler}>
                <CustomInput 
                    id='name' 
                    type='text' 
                    autoComplete="current-name" 
                    name='name' 
                    labelText='Full Name' 
                    value={name} 
                    required 
                    onChange={(e)=>setName(e.target.value)}/>
                <CustomInput 
                    id='email' 
                    type='email' 
                    autoComplete="current-email" 
                    name='email' 
                    labelText='Email Address' 
                    value={email} 
                    required 
                    onChange={(e)=>setEmail(e.target.value)}/>      
                <CustomInput 
                    id='password' 
                    type='password' 
                    autoComplete="new-password" 
                    name='password'
                    labelText='Password'
                    value={password} 
                    required 
                    onChange={(e)=>setPassword(e.target.value)}/>   
                <CustomInput
                    id='confirmPassword'
                    type='password'
                    autoComplete="new-password"
                    name='confirmPassword'
                    labelText='Confirm Password'
                    value={confirmPassword}
                    required
                    onChange={(e)=>setConfirmPassword(e.target.value)}/>
                    <Button className="my-2" type='submit' variant="primary">
                        Update
                    </Button>    
                </Form>
            </Col>
            <Col lg={8}>
                <h2>My Orders</h2>
                {loadingOrders?(<Loader/>):errorOrders?(<Message variant='danger'>{errorOrders}</Message>):(
                    <Table striped bordered hover responsive className='table-sm align-items-center' style={{textAlign:'center'}}>
                        <thead>
                            <tr>
                                <th>ID</th>
                                <th>DATE</th>
                                <th>TOTAL</th>
                                <th>PAID</th>
                                <th>DELIVERED</th>
                                <th>ACTIONS</th>
                            </tr>
                        </thead>
                        <tbody>
                            {orders.map(order=>(
                                <tr key={order._id}>
                                    <td>{order._id}</td>
                                    <td>{order.createdAt.substring(0,10)}</td>
                                    <td>{order.totalPrice.toFixed(2)}</td>
                                    <td>{order.isPaid?order.paidAt.substring(0,10):(<i className="fa-solid fa-circle-xmark" style={{color:'red'}}></i>)}</td>
                                    <td>{order.isDelivered?order.deliveredAt.substring(0,10):(<i className="fa-solid fa-circle-xmark" style={{color:'red'}}></i>)}</td>
                                    <td><LinkContainer to={`/order/${order._id}`}><Button variant='light' className="btn-sm">Order Details</Button></LinkContainer></td>
                                </tr>
                            ))}
                        </tbody>
                    </Table>
                )}
            </Col>
        </Row>
    </div>
  );
};

export default Profile;