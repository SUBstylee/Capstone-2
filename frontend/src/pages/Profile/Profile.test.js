import { render } from '@testing-library/react';
import Profile from './Profile.jsx';
import store from '../../store';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

test("renders ProductDetails component without crashing", () => {
    const { getByText } = render(
        <Provider store={store}><BrowserRouter><Profile /></BrowserRouter></Provider>
    );

    expect(getByText("User Profile")).toBeInTheDocument();
});

test("matches snapshot", function () {
    const { asFragment } = render(
        <Provider store={store}><BrowserRouter><Profile /></BrowserRouter></Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});