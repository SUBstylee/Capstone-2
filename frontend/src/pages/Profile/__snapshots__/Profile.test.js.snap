// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Profile"
  >
    <div
      class="row"
    >
      <div
        class="col-lg-4"
      >
        <h2>
          User Profile
        </h2>
        <form
          class=""
        >
          <div
            class="CustomInput"
          >
            <input
              autocomplete="current-name"
              id="name"
              name="name"
              required=""
              type="text"
              value=""
            />
            <label
              for="name"
            >
              <span
                style="transition-delay: 0ms;"
              >
                F
              </span>
              <span
                style="transition-delay: 50ms;"
              >
                u
              </span>
              <span
                style="transition-delay: 100ms;"
              >
                l
              </span>
              <span
                style="transition-delay: 150ms;"
              >
                l
              </span>
              <span
                style="transition-delay: 200ms;"
              >
                 
              </span>
              <span
                style="transition-delay: 250ms;"
              >
                N
              </span>
              <span
                style="transition-delay: 300ms;"
              >
                a
              </span>
              <span
                style="transition-delay: 350ms;"
              >
                m
              </span>
              <span
                style="transition-delay: 400ms;"
              >
                e
              </span>
            </label>
          </div>
          <div
            class="CustomInput"
          >
            <input
              autocomplete="current-email"
              id="email"
              name="email"
              required=""
              type="email"
              value=""
            />
            <label
              for="email"
            >
              <span
                style="transition-delay: 0ms;"
              >
                E
              </span>
              <span
                style="transition-delay: 50ms;"
              >
                m
              </span>
              <span
                style="transition-delay: 100ms;"
              >
                a
              </span>
              <span
                style="transition-delay: 150ms;"
              >
                i
              </span>
              <span
                style="transition-delay: 200ms;"
              >
                l
              </span>
              <span
                style="transition-delay: 250ms;"
              >
                 
              </span>
              <span
                style="transition-delay: 300ms;"
              >
                A
              </span>
              <span
                style="transition-delay: 350ms;"
              >
                d
              </span>
              <span
                style="transition-delay: 400ms;"
              >
                d
              </span>
              <span
                style="transition-delay: 450ms;"
              >
                r
              </span>
              <span
                style="transition-delay: 500ms;"
              >
                e
              </span>
              <span
                style="transition-delay: 550ms;"
              >
                s
              </span>
              <span
                style="transition-delay: 600ms;"
              >
                s
              </span>
            </label>
          </div>
          <div
            class="CustomInput"
          >
            <input
              autocomplete="new-password"
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
            <label
              for="password"
            >
              <span
                style="transition-delay: 0ms;"
              >
                P
              </span>
              <span
                style="transition-delay: 50ms;"
              >
                a
              </span>
              <span
                style="transition-delay: 100ms;"
              >
                s
              </span>
              <span
                style="transition-delay: 150ms;"
              >
                s
              </span>
              <span
                style="transition-delay: 200ms;"
              >
                w
              </span>
              <span
                style="transition-delay: 250ms;"
              >
                o
              </span>
              <span
                style="transition-delay: 300ms;"
              >
                r
              </span>
              <span
                style="transition-delay: 350ms;"
              >
                d
              </span>
            </label>
          </div>
          <div
            class="CustomInput"
          >
            <input
              autocomplete="new-password"
              id="confirmPassword"
              name="confirmPassword"
              required=""
              type="password"
              value=""
            />
            <label
              for="confirmPassword"
            >
              <span
                style="transition-delay: 0ms;"
              >
                C
              </span>
              <span
                style="transition-delay: 50ms;"
              >
                o
              </span>
              <span
                style="transition-delay: 100ms;"
              >
                n
              </span>
              <span
                style="transition-delay: 150ms;"
              >
                f
              </span>
              <span
                style="transition-delay: 200ms;"
              >
                i
              </span>
              <span
                style="transition-delay: 250ms;"
              >
                r
              </span>
              <span
                style="transition-delay: 300ms;"
              >
                m
              </span>
              <span
                style="transition-delay: 350ms;"
              >
                 
              </span>
              <span
                style="transition-delay: 400ms;"
              >
                P
              </span>
              <span
                style="transition-delay: 450ms;"
              >
                a
              </span>
              <span
                style="transition-delay: 500ms;"
              >
                s
              </span>
              <span
                style="transition-delay: 550ms;"
              >
                s
              </span>
              <span
                style="transition-delay: 600ms;"
              >
                w
              </span>
              <span
                style="transition-delay: 650ms;"
              >
                o
              </span>
              <span
                style="transition-delay: 700ms;"
              >
                r
              </span>
              <span
                style="transition-delay: 750ms;"
              >
                d
              </span>
            </label>
          </div>
          <button
            class="my-2 btn btn-primary"
            type="submit"
          >
            Update
          </button>
        </form>
      </div>
      <div
        class="col-lg-8"
      >
        <h2>
          My Orders
        </h2>
        <div
          class="table-responsive"
        >
          <table
            class="table-sm align-items-center table table-striped table-bordered table-hover"
            style="text-align: center;"
          >
            <thead>
              <tr>
                <th>
                  ID
                </th>
                <th>
                  DATE
                </th>
                <th>
                  TOTAL
                </th>
                <th>
                  PAID
                </th>
                <th>
                  DELIVERED
                </th>
                <th>
                  ACTIONS
                </th>
              </tr>
            </thead>
            <tbody />
          </table>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
